### ISSUE TITLE
non-default hooks?

### ISSUE THREAD

**Issue by tmccombs** [mawww/kakoune#3542]
In the documentation for hooks the hooks are listed under a section titled "Default hooks". Are there non-default hooks?

After thinking about this a little bit, I realized that it is kind of possible to create custom hooks, by creating a dummy option, and then setting hook on *SetOption, and triggering the hooks by setting the option.

Furthermore, I'd like a way to trigger one of the default hooks, with a specific value. In particular, if I have a plugin for a template language, such as erb, then I'd like to be able to load the syntax highlighting for the buffer as if the template suffix wasn't on the filename. For example, if I edit `test.html.erb` I'd like to trigger the BufferCreate hooks for `test.html` to get any highlighting or special commands and keys for html in addition to erb sepecific things. It's possible to duplicate all the html hooks for *.html.erb of course, but since erb could be used for practically any type of file, that doesn't scale very well.

p.s. is there any documentation on the order hooks are executed in?

**Comment by Screwtapello** [comment]
I don't know why the heading says "Default hooks"; as far as I know there's no hooks other than the ones listed.

There's no way to manually trigger a specific hook, but if you just want a buffer to be treated as a specific file type, you can just `set-option window filetype html` - the Kakoune convention is to do all the file-type setup in a WinSetOption hook, specifically so that it can be triggered when necessary by other code.

**Comment by lenormf** [comment]
It's supposed to mean "Built-in". Feel free to create a PR, "Default" is indeed misleading.

**Comment by tmccombs** [comment]
> you can just `set-option window filetype html`

but what if I don't have an exhaustive list of filetypes that could be used? what if the user changes the filetype to something else? Ideally, I'd set the filetype to something like erb.html, but then it wouldn't match the hooks for the html plugin. If I use `filetype=html`, then if the filetype is changed, I won't know if the erb options should be disabled or not.

**Comment by mawww** [comment]
I think relying on hooks for the use case you describe could be a bit brittle.

The common way to have erb re-use existing html support would be to:

* `require-module html` to ensure it exists
* use a `ref` highlighter to pull the `html` highlighter (so manually `add-highlighter window/ ref html`, likely more complex than that to apply additional erb higlighting in the correct order)
* Manually add hooks to execute `html-indent-...`

In other words, rely on the modular part of html.kak, but not on the rest, as you'd likely already have to write those hooks to handle erb specific logic )else you could just use filetype html), and adding 5 lines to them re-use what html.kak provides would be an ok tradeoff to make erb/html interaction explicit and easy to predict.

**Comment by tmccombs** [comment]
> and adding 5 lines to them re-use what html.kak provides would be an ok tradeoff

5 lines times however many languages you might use erb with...

**Comment by Screwtapello** [comment]
> but what if I don't have an exhaustive list of filetypes that could be used?

Yeah, Kakoune doesn't have a good, general-purpose system for dealing with "overlaid" file-formats like erb and similar macro pre-processors. In fairness, it's not generally possible to combine two file-formats like that, especially when one of them can dynamically generate the other at runtime, so even a "good" system would still be brittle.

The easiest thing to do would just be to do erb highlighting in erb files, and ignore the content outside the `<% %>` bits. Easy, but not very satisfying.

A more complex thing would be to have hooks that match `.*\.html\.erb` and set `filetype=html` (triggering all the usual HTML config) and then manually adding a bunch of erb stuff on top. That would have to be done per-filetype, but would be pretty easy for end-users to write manually, especially if you provided an `add-erb-highlighting` helper command.

The most complex thing would be to do something like: when a buffer ending in `.*\.erb` is created, strip off the `.erb` extension and create a scratch buffer with the resulting name. After all the filetype-detection logic has run, figure out what *that* buffer's `filetype` option is set to, destroy the buffer, then set the original buffer's `filetype` to that value. After all the highlighting hooks have run, add your erb stuff on top as in the previous approach. The result would be quite complex, probably brittle, and a mess to debug, but it *would* be automatic.

**Comment by lenormf** [comment]
It's off-topic, but I think that's what the following highlighter script does:

https://github.com/kakoune-editor/kakoune-extra-filetypes/blob/master/rc/simple-template.kak
